### PR TITLE
[E2 - E3 - E3.1] - Cargo Service

### PR DESCRIPTION
**Pull request by rambodghx**
Hi @mohammadKarimi,

To implement the cargo-related use case has described earlier, this PR has been created.

Please note the following considerations:

- Validation has been ignored at this phase.
- The naming conventions used were sourced from Google. I apologize for any inaccuracies or unfamiliarity with this domain.

### CODE REVIEW

File: Factory.Cargo/TransportationFactory/AirTransportationFactory.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using Factory.Cargo.Factory;
+using Factory.Cargo.Transportation;
+
+namespace Factory.Cargo.TransportationFactory;
+
+public class AirTransportationFactory : ITransportationFactory
```
Comment (mohammadKarimi) [inline Factory.Cargo/TransportationFactory/AirTransportationFactory.cs:6]: where are you using this Factory ?
i didn't find it.

File: Factory.Cargo/CargoService.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+public class CargoService
+{
+    public ITransportation CreateTransportation(
+        TransportationMode transportationMode,
+        string? origin,
+        string? destination)
+    {
+        return transportationMode switch
+        {
+            TransportationMode.Air => new AirTransportation(),
```
Comment (mohammadKarimi) [inline Factory.Cargo/CargoService.cs:22]: Air must be single instance.

File: Factory.Cargo/TransportationFactory/AirTransportationFactory.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+{
+    private ITransportation? _transportation;
+
+    public ITransportation Create()
+        => _transportation ??= new AirTransportation();
```
Comment (mohammadKarimi) [inline Factory.Cargo/TransportationFactory/AirTransportationFactory.cs:11]: consider it is not thread safe, but for this case it is enough.

**Review by mohammadKarimi — COMMENTED** [review]
perfect one,

now move to E3.1

Imagine it is a library and we make it as a package. so users or customers, can't change your code, but they want to add a new behavior like Truck. so, how you can solve it.

File: Factory.Cargo/TransportationFactory/AirTransportationFactory.cs (hunk 3 above)
Comment (rambodghx) [reply Factory.Cargo/TransportationFactory/AirTransportationFactory.cs:11]: Thanks @mohammadKarimi, I think in main implementation we should use atomic operations like Interlock.CompareExchange.
What is your idea?

**Comment by rambodghx** [comment]
@mohammadKarimi please checkout my last commit **71b7714**.
I didn't use the factory instances and that was my mistake because, at first step I implemented the factory pattern and then added the **AbstractFactory** pattern so I just forgot make change on **CargoService**.
sorry about that I fixed this.

**Comment by rambodghx** [comment]
@mohammadKarimi your new ordered task , E3.1 , has been implemented in latest commit 84716ac.
Note the following considerations :
- This use-case have been implemented without DI pattern in order to comply the KISS rule.
- Validations have been ignored in this phase
- Enum have been changed to **String** because there is no inheritance concept within Enum.

I look forward for your review.

File: Factory.Cargo/CargoFactoryApi.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
+using Factory.Cargo.Factory;
+using Factory.Cargo.TransportationFactory;
+
+namespace Factory.Cargo;
+
+public class CargoFactoryApi
```
Comment (mohammadKarimi) [inline Factory.Cargo/CargoFactoryApi.cs:6]: why did you add API? 
it is  a bad naming selection.

File: Factory.Cargo/CargoFactoryApi.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+{
+    private Dictionary<string, ITransportationFactory> _transportationFactories = new();
+
+    public void AddBaseFactories()
```
Comment (mohammadKarimi) [inline Factory.Cargo/CargoFactoryApi.cs:10]: why you did not move it into constructor? 
you user must know to call this method before every usage.

File: Factory.Cargo/CargoFactoryApi.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        _transportationFactories.Add("Air",new AirTransportationFactory());
+        _transportationFactories.Add("Train",new TrainTransportationFactory());
+        _transportationFactories.Add("Ship",new ShipTransportationFactory());
+    }
+    
+    
+    public void AddTransportationFactory(string mode, ITransportationFactory transportationFactory)
+        => _transportationFactories.Add(mode, transportationFactory);
+
+    public ITransportation Create(string mode, string? origin = null, string? destination = null)
```
Comment (mohammadKarimi) [inline Factory.Cargo/CargoFactoryApi.cs:21]: if you have 10 factories and each of them needs different parameter. so how to solve it ? are you going to add 10 different null parameter?

File: Factory.Cargo/CargoFactoryApi.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        if (_transportationFactories.TryGetValue(mode, out var transportationFactory) || transportationFactory is null)
+            throw new NotImplementedException($"The factory {mode} didn't define.");
+        try
+        {
+            return transportationFactory.Create(origin, destination);
```
Comment (mohammadKarimi) [inline Factory.Cargo/CargoFactoryApi.cs:27]: if you have 10 differ factory. how to call them? 

in this case you are calling in a try catch and then call another in line 34.

**Review by mohammadKarimi — CHANGES_REQUESTED** [review]

File: Factory.Cargo/TransportationFactory/AirTransportationFactory.cs (hunk 1 above)
Comment (rambodghx) [reply Factory.Cargo/TransportationFactory/AirTransportationFactory.cs:6]: Fixed.

File: Factory.Cargo/CargoService.cs (hunk 2 above)
Comment (rambodghx) [reply Factory.Cargo/CargoService.cs]: Fixed.

File: Factory.Cargo/CargoFactoryApi.cs (hunk 4 above)
Comment (rambodghx) [reply Factory.Cargo/CargoFactoryApi.cs]: @mohammadKarimi
I was looking for some distinguish between package class and last practice class.
I have changed the name from Api to Service.
Please checkout [024ad3e].

File: Factory.Cargo/CargoFactoryApi.cs (hunk 5 above)
Comment (rambodghx) [reply Factory.Cargo/CargoFactoryApi.cs]: I have enabled user to figure out this feature by optional flag on Ctor.
Please checkout [fb9738f].

File: Factory.Cargo/CargoFactoryApi.cs (hunk 6 above)
Comment (rambodghx) [reply Factory.Cargo/CargoFactoryApi.cs]: @mohammadKarimi 
I've developed new way to pass parameters dynamically.
`
public class CargoFactoryService
{
 public ITransportation Create(Dictionary<string,object> parameters);
}
`

so the use now can call factory in this way
` public LibraryUser()
    {
        _factoryService.AddTransportationFactory("Truck", new TruckTransportationFactory());

        var truckTransportation = _factoryService.Create("Truck", new Dictionary<string, object>
        {
            { "TruckType", "Trailer" },
        });
    }`

File: Factory.Cargo/CargoFactoryApi.cs (hunk 7 above)
Comment (rambodghx) [reply Factory.Cargo/CargoFactoryApi.cs]: @mohammadKarimi 
this issue addressed on previous review comment.

**Review by rambodghx — COMMENTED** [review]
@mohammadKarimi 
Please checkout comments that I've left on your review including new commits.
Thanks.
